sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2747,6 +2747,17 @@ class Dataset[T] private[sql](
     }
   }
 
+  /**
+   * Collect a Dataset as ArrowPayload byte arrays and serve to PySpark.
+   */
+  private[sql] def collectAsArrowToPython(): Int = {
+    val payloadRdd = toArrowPayloadBytes()
+    val payloadByteArrays = payloadRdd.collect()
```

Comment:
You can stream out payloads as they come into the driver (maybe this is already happening). We may be able to play with the StreamWriter to reduce the driver memory usage in a follow up patch